src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,369 @@
+							     /*config=*/NULL, /*config_size=*/0);
+
+	comp_dbg(dev, "google_ctc_audio_processing_init(): Ready");
+
+	return 0;
+fail:
+	comp_err(dev, "google_ctc_audio_processing_init(): Failed");
+	if (cd) {
+		rfree(cd->input);
+		rfree(cd->output);
+		if (cd->state)
+			GoogleCtcAudioProcessingFree(cd->state);
+		rfree(cd);
+	}
+	rfree(dev);
+	return ret;
+}
+
+static int google_ctc_audio_processing_free(struct processing_module *mod)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	GoogleCtcAudioProcessingFree(cd->state);
+	cd->state = NULL;
+	rfree(cd->input);
+	rfree(cd->output);
+	rfree(cd);
+	rfree(dev);
+	return 0;
+}
+
+static int google_ctc_audio_processing_reconfigure(struct processing_module *mod)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	uint8_t *config;
+	size_t size;
+	int ret;
+
+	comp_dbg(dev, "google_ctc_audio_processing_reconfigure()");
+
+	if (!comp_is_current_data_blob_valid(cd->tuning_handler) &&
+	    !comp_is_new_data_blob_available(cd->tuning_handler)) {
+		/*
+		 * The data blob hasn't been available once so far.
+		 *
+		 * This looks redundant since the same check will be done in
+		 * comp_get_data_blob() below. But without this early return,
+		 * hundreds of warn message lines are produced per second by
+		 * comp_get_data_blob() calls until the data blob is arrived.
+		 */
+		return 0;
+	}
+
+	config = comp_get_data_blob(cd->tuning_handler, &size, NULL);
+	if (size == 0) {
+		/* No data to be handled */
+		return 0;
+	}
+
+	if (!config) {
+		comp_err(dev, "google_ctc_audio_processing_reconfigure(): Tuning config not set");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
+		  config, size);
+
+	cd->reconfigure = false;
+
+	uint8_t *google_ctc_audio_processing_config;
+	size_t google_ctc_audio_processing_config_size;
+	int num_channels;
+	bool google_ctc_audio_processing_config_present;
+	bool num_channels_present;
+
+	GoogleCtcAudioProcessingParseSofConfigMessage(config, size,
+						      &google_ctc_audio_processing_config,
+						      &google_ctc_audio_processing_config_size,
+						      &num_channels,
+						      &google_ctc_audio_processing_config_present,
+						      &num_channels_present);
+
+	if (google_ctc_audio_processing_config_present) {
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying config of size %zu bytes",
+			  google_ctc_audio_processing_config_size);
+
+		ret = GoogleCtcAudioProcessingReconfigure(cd->state,
+							  google_ctc_audio_processing_config,
+							  google_ctc_audio_processing_config_size);
+		if (ret) {
+			comp_err(dev, "GoogleCtcAudioProcessingReconfigure failed: %d",
+				 ret);
+			return ret;
+		}
+	}
+
+	if (num_channels_present) {
+		cd->num_channels = num_channels;
```

Comment:
Why is num_channels an external tuneable at all?  Channel count is a runtime property of the SOF stream, shouldn't we be configuring this automatically to eliminate the possibility of getting wrong all the way out in Linux userspace (where it's a static binary blob that can't be programmatically controlled except at image generation time, no less!)?

Maybe it's too late to fix the external library API.  But at the very least I'd suggest validating this vs. the "real" value we get from the SOF input source and using that one if there's a conflict.

Reply:
Yes, `num_channels` should not be an tuneable value. I'll remove it. Excuse me for the confusion.